### PR TITLE
[PETSc] version 3.19.6

### PR DESCRIPTION
**Pull request by boriskaus**
- bump version to 3.19.6
- simplify dependencies, that hopefully reduce runtime errors moving forward

### CODE REVIEW

**Comment by JordiManyer** [comment]
Hey @boriskaus any luck with getting this PR running?

**Comment by boriskaus** [comment]
Sorry this slipped my attention- hope to get back to it next week

**Comment by boriskaus** [comment]
All compile fine now, apart from these two:
`aarch64-linux-gnu-libgfortran3` 
`x86-64-linux-musl-libgfortran5`

They weirdly have a segmentation fault during compilation, see [here](https://buildkite.com/julialang/yggdrasil/builds/7497#018ccb78-5c86-48fa-8465-c31b1f94e0af/6-9394) and [here](https://buildkite.com/julialang/yggdrasil/builds/7497#018ccb78-5c86-48fa-8465-c31b1f94e0af/6-9393):

```
[19:22:00] cc -c -fno-stack-protector -fPIC -O3  -I/workspace/srcdir/petsc-3.19.6/include -I/workspace/srcdir/petsc-3.19.6/aarch64-linux-gnu_double_real_Int64/include -I/workspace/destdir/include  -I/workspace/destdir/include  -MMD -MP /workspace/srcdir/petsc-3.19.6/src/mat/impls/baij/seq/baijsolvnat4.c -o aarch64-linux-gnu_double_real_Int64/obj/mat/impls/baij/seq/baijsolvnat4.o
--
  | [19:22:00] /workspace/srcdir/petsc-3.19.6/src/mat/impls/baij/seq/baijsolvnat4.c: In function ‘MatSolve_SeqBAIJ_4_NaturalOrdering_inplace’:
  | [19:22:00] /workspace/srcdir/petsc-3.19.6/src/mat/impls/baij/seq/baijsolvnat4.c:105:1: internal compiler error: Segmentation fault
  | [19:22:00]  }
  | [19:22:00]  ^
  | [19:22:00] 0x9a3336 crash_signal
  | [19:22:00] 	/workspace/srcdir/gcc-6.1.0/gcc/toplev.c:333
  | [19:22:00] 0x7f8d2479a1de ???
  | [19:22:00] 	/workspace/srcdir/musl-1.2.2/src/signal/x86_64/restore.s:1
  | [19:22:00] Please submit a full bug report,
  | [19:22:00] with preprocessed source if appropriate.
  | [19:22:00] Please include the complete backtrace with any bug report.
  | [19:22:00] See <http://gcc.gnu.org/bugs.html> for instructions.
``` 
Weirdly, this doesn't happen with the same architecture and a different compiler version. I can reproduce the issue locally, but I'm not making any progress in resolving it.
Easiest is to deactivate the two problematic ones.

**Comment by ViralBShah** [comment]
Should we try to get a newer version of petsc in?

**Comment by boriskaus** [comment]
I can give it a go after I’m back from holidays (few days). Yet we do need a more maintainable way to deal with the external dependencies. 
You suggested earlier to compile suitesparse as a separate, compile-time, library. Is it possible to compile it as a static library and link it with petsc at compile time? 
If yes, how is this done?

**Comment by ViralBShah** [comment]
If Petsc has moved to SuiteSparse 7, I believe the APIs are fairly stable and it should just work. Shall we take a look at that option first?

**Comment by boriskaus** [comment]
Seems to be using 7.8.0 now, see [here](https://gitlab.com/petsc/petsc/-/blob/main/config/BuildSystem/config/packages/SuiteSparse.py#L9)

**Comment by ViralBShah** [comment]
We are on 7.7. I can easily bump to 7.8 - but I am pretty sure the APIs are compatible.

https://github.com/JuliaLang/julia/blob/master/deps/libsuitesparse.version

**Comment by boriskaus** [comment]
Ok worth a shot - I’ll try later this week

**Comment by boriskaus** [comment]
ok, I managed to compile a version of PETSc 3.21.4 with LBT, while adding the `_x64` suffix, you suggested [earlier](https://github.com/JuliaPackaging/Yggdrasil/pull/7989#issuecomment-1906918827). At least for Mac, the library compiles but gives the following mistake when trying to test it.

```Julia
(test_PETSc_jll) pkg> test
     Testing test_PETSc_jll
      Status `/private/var/folders/8g/m1k97pb96r76xpr3n9k947rh0000gr/T/jl_PEJq5y/Project.toml`
  [8fa3689e] PETSc_jll v3.21.4+0 `https://github.com/boriskaus/PETSc_jll.jl#main`
  [6087adde] test_PETSc_jll v0.1.0 `~/.julia/dev/test_PETSc_jll`
  [44cfe95a] Pkg v1.10.0
  [8dfed614] Test
      Status `/private/var/folders/8g/m1k97pb96r76xpr3n9k947rh0000gr/T/jl_PEJq5y/Manifest.toml`
  [692b3bcd] JLLWrappers v1.5.0
  [3da0fdf6] MPIPreferences v0.1.11
  [21216c6a] Preferences v1.4.3
  [6e34b625] Bzip2_jll v1.0.8+1
  [e33a78d0] Hwloc_jll v2.11.1+0
  [1d63c593] LLVMOpenMP_jll v15.0.7+0
  [d00139f3] METIS_jll v5.1.2+0
  [7cb0a576] MPICH_jll v4.2.2+0
  [f1f71cc9] MPItrampoline_jll v5.4.0+0
  [ca64183c] MUMPS_jll v5.7.3+0
  [9237b28f] MicrosoftMPI_jll v10.1.4+2
⌅ [656ef2d0] OpenBLAS32_jll v0.3.24+0
  [fe0851c0] OpenMPI_jll v5.0.5+0
  [b247a4be] PARMETIS_jll v4.0.6+2
  [8fa3689e] PETSc_jll v3.21.4+0 `https://github.com/boriskaus/PETSc_jll.jl#main`
  [b3ec0f5a] PTSCOTCH_jll v7.0.4+0
  [aabda75e] SCALAPACK32_jll v2.2.1+1
  [a8d0f55d] SCOTCH_jll v7.0.4+3
  [9a1356b0] SuperLU_DIST_jll v8.2.1+0
  [ffd25f8a] XZ_jll v5.4.6+0
  [6087adde] test_PETSc_jll v0.1.0 `~/.julia/dev/test_PETSc_jll`
  [0dad84c5] ArgTools v1.1.1
  [56f22d72] Artifacts
  [2a0f44e3] Base64
  [ade2ca70] Dates
  [f43a241f] Downloads v1.6.0
  [7b1f6079] FileWatching
  [b77e0a4c] InteractiveUtils
  [4af54fe1] LazyArtifacts
  [b27032c2] LibCURL v0.6.4
  [76f85450] LibGit2
  [8f399da3] Libdl
  [56ddb016] Logging
  [d6f4376e] Markdown
  [ca575930] NetworkOptions v1.2.0
  [44cfe95a] Pkg v1.10.0
  [de0858da] Printf
  [3fa0cd96] REPL
  [9a3f8284] Random
  [ea8e919c] SHA v0.7.0
  [9e88b42a] Serialization
  [6462fe0b] Sockets
  [fa267f1f] TOML v1.0.3
  [a4e569a6] Tar v1.10.0
  [8dfed614] Test
  [cf7118a7] UUIDs
  [4ec0a83e] Unicode
  [e66e0078] CompilerSupportLibraries_jll v1.1.1+0
  [deac9b47] LibCURL_jll v8.4.0+0
  [e37daf67] LibGit2_jll v1.6.4+0
  [29816b5a] LibSSH2_jll v1.11.0+1
  [c8ffd9c3] MbedTLS_jll v2.28.2+1
  [14a3606d] MozillaCACerts_jll v2023.1.10
  [83775a58] Zlib_jll v1.2.13+1
  [8e850b90] libblastrampoline_jll v5.8.0+1
  [8e850ede] nghttp2_jll v1.52.0+1
  [3f19e933] p7zip_jll v17.4.0+2
        Info Packages marked with ⌅ have new versions available but compatibility constraints restrict them from upgrading.
     Testing Running tests...
    Updating git-repo `https://github.com/boriskaus/PETSc_jll.jl`
   Resolving package versions...
  No Changes to `/private/var/folders/8g/m1k97pb96r76xpr3n9k947rh0000gr/T/jl_PEJq5y/Project.toml`
  No Changes to `/private/var/folders/8g/m1k97pb96r76xpr3n9k947rh0000gr/T/jl_PEJq5y/Manifest.toml`
ERROR: LoadError: InitError: could not load library "/Users/kausb/.julia/artifacts/5fb76d673cb21524f7bc2edf7d5d0d6910ce0ee1/lib/petsc/double_real_Int64/lib/libpetsc_double_real_Int64.3.21.4.dylib"
dlopen(/Users/kausb/.julia/artifacts/5fb76d673cb21524f7bc2edf7d5d0d6910ce0ee1/lib/petsc/double_real_Int64/lib/libpetsc_double_real_Int64.3.21.4.dylib, 0x0001): symbol not found in flat namespace '_descinit_64_'
Stacktrace:
  [1] dlopen(s::String, flags::UInt32; throw_error::Bool)
    @ Base.Libc.Libdl ./libdl.jl:117
  [2] dlopen(s::String, flags::UInt32)
    @ Base.Libc.Libdl ./libdl.jl:116
  [3] macro expansion
    @ ~/.julia/packages/JLLWrappers/pG9bm/src/products/library_generators.jl:63 [inlined]
  [4] __init__()
    @ PETSc_jll ~/.julia/packages/PETSc_jll/qtxdl/src/wrappers/aarch64-apple-darwin-libgfortran5-mpi+mpich.jl:24
  [5] run_module_init(mod::Module, i::Int64)
    @ Base ./loading.jl:1134
  [6] register_restored_modules(sv::Core.SimpleVector, pkg::Base.PkgId, path::String)
    @ Base ./loading.jl:1122
  [7] _include_from_serialized(pkg::Base.PkgId, path::String, ocachepath::String, depmods::Vector{Any})
    @ Base ./loading.jl:1067
  [8] _require_search_from_serialized(pkg::Base.PkgId, sourcepath::String, build_id::UInt128)
    @ Base ./loading.jl:1581
  [9] _require(pkg::Base.PkgId, env::String)
    @ Base ./loading.jl:1938
 [10] __require_prelocked(uuidkey::Base.PkgId, env::String)
    @ Base ./loading.jl:1812
 [11] #invoke_in_world#3
    @ ./essentials.jl:926 [inlined]
 [12] invoke_in_world
    @ ./essentials.jl:923 [inlined]
 [13] _require_prelocked(uuidkey::Base.PkgId, env::String)
    @ Base ./loading.jl:1803
 [14] macro expansion
    @ ./loading.jl:1790 [inlined]
 [15] macro expansion
    @ ./lock.jl:267 [inlined]
 [16] __require(into::Module, mod::Symbol)
    @ Base ./loading.jl:1753
 [17] #invoke_in_world#3
    @ ./essentials.jl:926 [inlined]
 [18] invoke_in_world
    @ ./essentials.jl:923 [inlined]
 [19] require(into::Module, mod::Symbol)
    @ Base ./loading.jl:1746
 [20] include(fname::String)
    @ Base.MainInclude ./client.jl:489
 [21] top-level scope
    @ none:6
during initialization of module PETSc_jll
in expression starting at /Users/kausb/.julia/dev/test_PETSc_jll/test/runtests.jl:8
ERROR: Package test_PETSc_jll errored during testing
```

Any suggestions?

**Comment by giordano** [comment]
Isn't the suffix `64_`, not `_64`?  Also, is `descinit` a BLAS function?

**Comment by boriskaus** [comment]
if using `x86_` the configuration fails with:
```
Executing: cc  -o /tmp/petsc-0td53c1p/config.libraries/conftest -L/workspace/destdir/lib -L/workspace/destdir/lib/darwin -lclang_rt.osx -Wl,-bind_at_load -Wl,-commons,use_dylibs -Wl,-search_paths_first -Wl,-no_compact_unwind -fno-stack-protector -fPIC -O3 /tmp/petsc-0td53c1p/config.libraries/conftest.o  -Wl,-rpath,/workspace/destdir/lib -L/workspace/destdir/lib -lblastrampoline -Wl,-rpath,/workspace/destdir/lib -L/workspace/destdir/lib -lmpifort -Wl,-rpath,/workspace/destdir/lib -L/workspace/destdir/lib -lmpi -lgfortran  -lgcc_s.1
stdout:
Undefined symbols for architecture arm64:
  "_ddot64_", referenced from:
      _main in conftest.o
ld: symbol(s) not found for architecture arm64
clang-16: error: linker command failed with exit code 1 (use -v to see invocation)
Possible ERROR while running linker: exit code 1
```
`descinit` is indeed a SCALAPACK function, and not BLAS. Perhaps PETSc propagates the suffix onwards?

**Comment by ViralBShah** [comment]
The symbol should be `_ddot_64_`.

**Comment by boriskaus** [comment]
ok, I locally compiled a version of PETSc 3.21.4 with only LBT as dependency (removing all other external packages). It compiles fine, but if I try to run it, it crashes, with the error below (see also [here](https://github.com/boriskaus/test_PETSc_jll/actions/runs/10401906966/job/28805313735#step:7:159)):

```
lid velocity = 0.0016, prandtl # = 1., grashof # = 1.
[0]PETSC ERROR: ------------------------------------------------------------------------
[0]PETSC ERROR: Caught signal number 11 SEGV: Segmentation Violation, probably memory access out of range
[0]PETSC ERROR: Try option -start_in_debugger or -on_error_attach_debugger
[0]PETSC ERROR: or see https://petsc.org/release/faq/#valgrind and https://petsc.org/release/faq/
[0]PETSC ERROR: ---------------------  Stack Frames ------------------------------------
[0]PETSC ERROR: The line numbers in the error traceback are not always exact.
[0]PETSC ERROR: #1 BLASdot()
[0]PETSC ERROR: #2 VecNorm_Seq() at /workspace/srcdir/petsc-3.21.4/src/vec/vec/impls/seq/bvec2.c:206
[0]PETSC ERROR: #3 VecNorm() at /workspace/srcdir/petsc-3.21.4/src/vec/vec/interface/rvector.c:226
[0]PETSC ERROR: #4 SNESSolve_NEWTONLS() at /workspace/srcdir/petsc-3.21.4/src/snes/impls/ls/ls.c:173
[0]PETSC ERROR: #5 SNESSolve() at /workspace/srcdir/petsc-3.21.4/src/snes/interface/snes.c:4753
[0]PETSC ERROR: #6 main() at ex19.c:152
application called MPI_Abort(MPI_COMM_WORLD, 59) - process 0
```

within PETSc, the [offending line](https://gitlab.com/petsc/petsc/-/blob/main/src/vec/vec/impls/seq/bvec2.c?ref_type=heads#L206) is:
```
  PetscCallBLAS("BLASdot", ztmp[type == NORM_1_AND_2] = PetscSqrtReal(PetscRealPart(BLASdot_(&bn, xx, &one, xx, &one))));
``` 

any idea?

**Comment by ViralBShah** [comment]
cc @staticfloat @giordano Thoguhts?

**Comment by boriskaus** [comment]
ok, I made some progress by doing 2 changes:

1) I initially did not link the final executable with a local BLAS implementation. I added this (taken from the LBT testing framework):

```julia
function add_LBT_flags(cmd::Cmd)
    # using LBT requires to set the following environment variables
    libdirs = unique(vcat(OpenBLAS32_jll.LIBPATH_list..., CompilerSupportLibraries_jll.LIBPATH_list...))
    backing_libs = OpenBLAS32_jll.libopenblas_path
    
    env = Dict(
        # We need to tell it how to find CSL at run-time
        LIBPATH_env => append_libpath(libdirs, cmd.env),
        "LBT_DEFAULT_LIBS" => backing_libs,
        "LBT_STRICT" => 1,
        "LBT_VERBOSE" => 1,
    )
    cmd = addenv(cmd, env)
    
    return cmd
end

 cmd = add_LBT_flags(cmd)
``` 
using this, the CI [works](https://github.com/boriskaus/test_PETSc_jll/actions/runs/10405187540/job/28815380724) on Mac and linux for various Julia versions except for windows. 

2) To get that working, I had to undo the `_64` option again.
 
any ideas on how to make LBT work on Windows?

**Comment by ViralBShah** [comment]
In some cases I used `libblastrampoline-5` on Windows. I don't know if that is still necessary.

**Comment by boriskaus** [comment]
fwiw, I'm currently linking it as follows:
```Julia
    if [[ "${target}" == aarch64-apple-* ]]; then    
        LIBFLAGS="-L${libdir}" 
        # Linking requires the function `__divdc3`, which is implemented in
        # `libclang_rt.osx.a` from LLVM compiler-rt.
        BLAS_LAPACK_LIB="${libdir}/libblastrampoline.${dlext}"
        CLINK_FLAGS="-L${libdir}/darwin -lclang_rt.osx"
    elif [[ "${target}" == *-mingw* ]]; then
        BLAS_LAPACK_LIB="${libdir}/libblastrampoline-5.${dlext}"
        CLINK_FLAGS=""
    else
        BLAS_LAPACK_LIB="${libdir}/libblastrampoline.${dlext}"
        CLINK_FLAGS=""
    fi
```

**Comment by staticfloat** [comment]
@boriskaus can you give me a simplified recipe to reproduce the issues you're seeing?  The CI setup looks more complicated than I expected.

**Comment by boriskaus** [comment]
fair point. I've added the [buildscript.jl](https://github.com/boriskaus/test_PETSc_jll/blob/main/build_scripts/build_tarballs_3.21.4_LBT.jl) and a minimum [example](https://github.com/boriskaus/test_PETSc_jll/blob/main/test/test_MWE.jl) with/without MPI to the repo. Hope that is easier to work with.

**Comment by boriskaus** [comment]
@staticfloat: is `libblastrampoline` working on windows? If yes, how do I set the environmental variables such that the executable in the shell know which `BLAS` library to call? 
I didn't find a documentation related to that and the CI of `libblastrampoline` doesn't seem to include windows, as far as I can tell.

**Comment by ViralBShah** [comment]
Yes LBT works on windows. We have been shipping it by default for several releases.

**Comment by boriskaus** [comment]
if I use OpenBLAS on windows and LBT on the other systems, it [works](https://github.com/boriskaus/test_PETSc_jll/actions/runs/10419200893) (still w/out external packages).

As I'm still bugged by the fact that MPI stopped working on Windows somewhere last year, I'm giving it a few more shots with older versions of `MSMPI`.

**Comment by boriskaus** [comment]
As an update, the following combination seems to work on PETSc 3.21.4:

- `SuiteSparse` on linux & mac, compiled as static library 
- `SuperLU_Dist` on linux and mac compiled statically
- `MUMPS` on linux compiled statically
- `MUMPS` on Mac as dynamic library
- MPI on windows remains broken, so this remains deactivated

On PETSc 3.19, `SuiteSparse` doesn't seem to work so that remains inactive.
As we also need PETSc versions 3.19 and 3.20, I'll deal with those before pushing 3.21

File: P/PETSc/build_tarballs.jl [hunk 1]
```diff
@@ -316,9 +365,24 @@ platforms = expand_gfortran_versions(supported_platforms(exclude=[Platform("i686
...
                                                                   Platform("armv7l","linux"; libc="musl"),
                                                                   Platform("armv7l","linux"; libc="gnu"),
-                                                                  Platform("aarch64","linux"; libc="musl")]))
-platforms, platform_dependencies = MPI.augment_platforms(platforms; MPItrampoline_compat=MPItrampoline_compat_version)
-
+                                                                  Platform("aarch64","linux"; libc="musl"),
+                                                                  Platform("aarch64","linux"; libc="musl"),
+                                                                  ]))
+
+# a few, but not all, platforms with libgfortran 3.0.0 are excluded
```
Comment (giordano) [inline P/PETSc/build_tarballs.jl:372]: I think it's easier to do filtering like https://github.com/JuliaPackaging/Yggdrasil/blob/7faddc5dca75e563154ad0cb5a1ab7ce84ed26fc/S/smesh/build_tarballs.jl#L37 Exact equality of platform objects is finicky, I don't recommend it. Basically you want to keep only freebsd and Windows with libgfortran 3? I fail to see a pattern here.

File: P/PETSc/build_tarballs.jl [hunk 2]
```diff
@@ -272,7 +296,7 @@ build_petsc()
     if  [ "${1}" == "double" ] &&  [ "${2}" == "real" ] &&  [ "${3}" == "Int64" ] &&  [ "${4}" == "deb" ]; then
         
         # this is the example that PETSc uses to test the correct installation        
-        # We compile it with debug flags (helpful to catch issues)
+        # We compile it witn debug flags (helpful to catch issues)
```
Comment (giordano) [inline P/PETSc/build_tarballs.jl:299]:
```suggestion
        # We compile it with debug flags (helpful to catch issues)
```

**Review by giordano — COMMENTED** [review]
Just to clarify: should we merge first this PR and then #9291?  Or #9291 supersedes this one?

**Comment by boriskaus** [comment]
Yes please first merge this one and then the others - PETSc is not backwards compatible so some codes need 3.19, whereas others can have 3.20 (and 3.21 which will come next).

**Pull request merged**